File: packages/compose/nockerl-design-components/src/main/kotlin/com/dizyx/nockerl/design/components/NockerlFacetedBackground.kt
package com.dizyx.nockerl.design.components

import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.produceState
import androidx.compose.runtime.withFrameNanos
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawWithCache
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.dizyx.nockerl.design.tokens.LocalNockerlColors
import kotlinx.coroutines.delay
import kotlin.math.PI
import kotlin.math.sin

// ── The Nockerl signature surface (package port) ───────────────────────
//
// A subtle, premium, GEOMETRIC field behind content: a low-poly mesh of large
// soft TRIANGULAR facets (on-brand, since the Nockerl mark is a triangle), each a
// whisper of tonal variation off the `chatBg` token, with a slow tone-wave
// drifting diagonally. NOT glow / aurora / neon; calm, "active but not busy".
//
// MAX-SMOOTHNESS TECHNIQUE (the native-butter contract):
// 1. The phase derives from a THROTTLED frame clock CAPPED at ~20fps (v1.14.0 perf
//    cut; was `rememberInfiniteTransition` running per-frame up to the display's
//    120fps). For the slow ~18s sine the per-frame luminance step is ~0.001
//    (sub-perceptual), so the look is IDENTICAL at a fraction of the CPU. `paused`
//    (host occlusion) or a non-animating host stops the clock outright: 0 CPU,
//    the static field stays on screen.
// 2. `drawWithCache` builds the triangulation ONCE per size; the per-frame work
//    is only a few dozen `drawPath` re-tints (no geometry rebuild, no
//    allocation), so the frame budget is trivially met and the drift never
//    stutters into visible steps.
// 3. All randomness is a DETERMINISTIC integer hash ([nockerlFacetJitter01]):
//    the field never re-rolls, so nothing shimmers between frames.
//
// Ported 1:1 from the canonical Android `ChatFeedBackground` (geometry, hash,
// constants); Swift renders the SAME field from the same pure functions.

/** Target facet cell size, kept LARGE so the mesh reads as a few big facets. */
private val FACET_CELL_SIZE: Dp = 128.dp

/** One full tone-wave cycle. ~18s reads alive, not busy. */
private const val WAVE_PERIOD_MILLIS: Int = 18_000

/**
 * Tone-wave sampling interval: ~20fps (v1.14.0 perf cut). The phase is re-derived
 * from the frame clock at most this often; for the ~18s sine the per-step luminance
 * change is ~0.001 (sub-perceptual), so 20fps looks identical at ~6x less CPU than
 * the display's up-to-120fps refresh.
 */
private const val FACET_FRAME_INTERVAL_MILLIS: Long = 50L

/** Peak luminance swing the moving tone-wave adds to a facet. */
private const val WAVE_AMPLITUDE: Float = 0.05f

/** Static per-facet luminance jitter range (± this fraction) baked at build. */
private const val STATIC_FACET_JITTER: Float = 0.022f

/** Alpha of the hairline drawn on each facet edge: barely-there structure. */
private const val EDGE_ALPHA: Float = 0.05f

/**
 * One triangular facet: three corners, a baked [staticDelta] tonal offset, and
 * its normalized [diagonalPos] (0..1 along the top-left → bottom-right
 * diagonal) that phases the travelling tone-wave.
 */
internal data class NockerlFacet(
    val a: Offset,
    val b: Offset,
    val c: Offset,
    val staticDelta: Float,
    val diagonalPos: Float,
)

/**
 * Deterministic pseudo-random in `0f..1f` from two integer grid indices: a
 * mixed-bit integer hash (splitmix-style avalanche), identical on Swift, so
 * both platforms render the SAME field and nothing ever re-rolls/shimmers.
 */
internal fun nockerlFacetJitter01(
    x: Int,
    y: Int,
): Float {
    var h = x * 374_761_393 + y * 668_265_263
    h = (h xor (h ushr 13)) * 1_274_126_177
    h = h xor (h ushr 16)
    return (h and 0xFFFFFF) / 16_777_216f
}

/**
 * Shift a color's luminance by [delta] (signed fraction of full scale),
 * clamping each channel; alpha preserved. Per-channel add keeps the facet hue
 * neutral (the field is tonal, not colored).
 */
internal fun nockerlShiftLuminance(
    color: Color,
    delta: Float,
): Color =
    Color(
        red = (color.red + delta).coerceIn(0f, 1f),
        green = (color.green + delta).coerceIn(0f, 1f),
        blue = (color.blue + delta).coerceIn(0f, 1f),
        alpha = color.alpha,
        colorSpace = color.colorSpace,
    )

/**
 * Total luminance delta for a facet at animation [phase] (radians): its baked
 * static delta plus `amplitude * sin(phase + diagonalPos * 2π)` (the sine
 * crest sweeps along the diagonal as the phase advances). Pure + frame-cheap.
 */
internal fun nockerlFacetLuminanceDelta(
    facet: NockerlFacet,
    phase: Float,
    amplitude: Float = WAVE_AMPLITUDE,
): Float = facet.staticDelta + amplitude * sin(phase + facet.diagonalPos * (2f * PI.toFloat()))

/**
 * Build the low-poly triangulation covering [size] with cells of [cellPx]:
 * a point grid (one extra ring beyond each edge so facets bleed off-screen),
 * interior points nudged by the deterministic hash, each cell split into two
 * triangles alternating the diagonal by parity. Pure + unit-testable.
 */
internal fun nockerlBuildFacetField(
    size: Size,
    cellPx: Float,
    jitterFraction: Float = 0.34f,
): List<NockerlFacet> {
    if (size.width <= 0f || size.height <= 0f || cellPx <= 0f) return emptyList()

    val cols = (size.width / cellPx).toInt() + 2
    val rows = (size.height / cellPx).toInt() + 2
    val jitterPx = cellPx * jitterFraction
    val diagSpan = (size.width + size.height).coerceAtLeast(1f)

    fun point(
        col: Int,
        row: Int,
    ): Offset {
        val baseX = col * cellPx
        val baseY = row * cellPx
        val edge = col <= 0 || row <= 0 || col >= cols - 1 || row >= rows - 1
        if (edge) return Offset(baseX, baseY)
        val dx = (nockerlFacetJitter01(col, row) - 0.5f) * 2f * jitterPx
        val dy = (nockerlFacetJitter01(col + 101, row + 211) - 0.5f) * 2f * jitterPx
        return Offset(baseX + dx, baseY + dy)
    }

    val facets = ArrayList<NockerlFacet>(cols * rows * 2)
    for (row in 0 until rows - 1) {
        for (col in 0 until cols - 1) {
            val tl = point(col, row)
            val tr = point(col + 1, row)
            val br = point(col + 1, row + 1)
            val bl = point(col, row + 1)
            val splitTlBr = (col + row) % 2 == 0
            val tri1 = if (splitTlBr) Triple(tl, tr, br) else Triple(tl, tr, bl)
            val tri2 = if (splitTlBr) Triple(tl, br, bl) else Triple(tr, br, bl)
            facets += facetOf(tri1, col, row, 0, diagSpan)
            facets += facetOf(tri2, col, row, 1, diagSpan)
        }
    }
    return facets
}

/** Assemble a facet: centroid diagonal position + hash-baked static delta. */
private fun facetOf(
    tri: Triple<Offset, Offset, Offset>,
    col: Int,
    row: Int,
    half: Int,
    diagSpan: Float,
): NockerlFacet {
    val (a, b, c) = tri
    val cx = (a.x + b.x + c.x) / 3f
    val cy = (a.y + b.y + c.y) / 3f
    val diagonalPos = ((cx + cy) / diagSpan).coerceIn(0f, 1f)
    val staticDelta = (nockerlFacetJitter01(col * 2 + half, row) - 0.5f) * 2f * STATIC_FACET_JITTER
    return NockerlFacet(a = a, b = b, c = c, staticDelta = staticDelta, diagonalPos = diagonalPos)
}

/** Triangle [Path] from a facet's corners (built once, reused per frame). */
private fun NockerlFacet.toPath(): Path =
    Path().apply {
        moveTo(a.x, a.y)
        lineTo(b.x, b.y)
        lineTo(c.x, c.y)
        close()
    }

/**
 * The **faceted background**, the Nockerl signature surface: an animated
 * low-poly triangle field filling its parent behind content (see the file
 * header for the geometry + the max-smoothness contract). Reads `chatBg` +
 * `canvasEdge` from [LocalNockerlColors], so it tracks the active palette.
 *
 * @param modifier layout modifier: typically `Modifier.matchParentSize()`.
 * @param animate when `true` (default) the tone-wave drifts continuously; when
 *   `false` the field is FROZEN at phase 0: for reduced-motion hosts and for
 *   DETERMINISTIC snapshots (the gallery/Roborazzi golden captures a static
 *   frame, since a running clock has no stable capture frame).
 * @param paused when `true` the tone-wave clock STOPS (0 CPU) while the static
 *   faceted field stays on screen: hosts set this when the window is occluded /
 *   backgrounded (v1.14.0). Composed with [animate]: the wave runs only when
 *   `animate && !paused`. Default `false`: the pre-1.14.0 behavior. Mirrors the
 *   Swift `paused` parameter.
 */
@Composable
fun NockerlFacetedBackground(
    modifier: Modifier = Modifier,
    animate: Boolean = true,
    paused: Boolean = false,
) {
    val colors = LocalNockerlColors.current
    val baseColor = colors.chatBg
    val edgeColor = colors.canvasEdge.copy(alpha = EDGE_ALPHA)
    val density: Density = LocalDensity.current
    val cellPx = with(density) { FACET_CELL_SIZE.toPx() }
    val edgeStroke = with(density) { Stroke(width = 0.75.dp.toPx()) }

    // Continuous, seamless phase 0 → 2π: a full sine cycle wraps cleanly (the
    // throttled clock loops modulo the period, no visible seam). Frozen at 0 when the
    // host isn't animating OR is paused (v1.14.0): the clock never composes/runs.
    val phase = if (animate && !paused) rememberFacetPhase() else 0f

    Spacer(
        modifier =
            modifier
                .fillMaxSize()
                .drawWithCache {
                    // Geometry: rebuilt ONLY when the draw size changes.
                    val paths = nockerlBuildFacetField(size, cellPx).map { it to it.toPath() }
                    onDrawBehind {
                        drawRect(color = baseColor)
                        paths.forEach { (facet, path) ->
                            val delta = nockerlFacetLuminanceDelta(facet, phase)
                            drawPath(path = path, color = nockerlShiftLuminance(baseColor, delta))
                            drawPath(path = path, color = edgeColor, style = edgeStroke)
                        }
                    }
                },
    )
}

/**
 * The drifting tone-wave phase (0 → 2π, ~18s, seamless loop), THROTTLED to ~20fps
 * (v1.14.0 perf cut). Isolated in its own composable so the public
 * [NockerlFacetedBackground] composes it ONLY when animating: a frozen / paused
 * field never runs the clock at all (0 CPU). The phase is derived from the frame
 * timestamp (like Swift's `TimelineView(.animation(minimumInterval:))`) and
 * re-sampled every [FACET_FRAME_INTERVAL_MILLIS]; `withFrameNanos` also parks the
 * loop whenever the host stops producing frames (occluded / backgrounded).
 */
@Composable
private fun rememberFacetPhase(): Float {
    val phase by produceState(initialValue = 0f) {
        val startNanos = withFrameNanos { it }
        while (true) {
            val nowNanos = withFrameNanos { it }
            val elapsedMillis = (nowNanos - startNanos) / 1_000_000f
            val cyclePos = (elapsedMillis % WAVE_PERIOD_MILLIS) / WAVE_PERIOD_MILLIS
            value = cyclePos * 2f * PI.toFloat()
            delay(FACET_FRAME_INTERVAL_MILLIS)
        }
    }
    return phase
}

File: Sources/NockerlDesign/Components/NockerlFacetedBackground.swift
// NockerlFacetedBackground: the Nockerl signature surface (package port).
//
// An animated low-poly triangle field behind content: large soft facets, each a
// whisper of tonal variation off the `chatBg` token, with a slow (~18s) tone-wave
// drifting diagonally. Calm, geometric, on-brand (the mark is a triangle). Never
// glow/aurora.
//
// MAX-SMOOTHNESS TECHNIQUE (the native-butter contract):
// 1. `TimelineView(.animation)`: CAPPED at 20fps via `minimumInterval` (v1.14.0
//    perf cut; was uncapped up to the display's 120fps, ~55% of a CPU core idle).
//    For the slow ~18s sine the per-frame luminance step is ~0.001 (sub-perceptual),
//    so the look is IDENTICAL at a fraction of the CPU. The phase still derives
//    continuously from the frame timestamp, nothing ticks; `paused` (or Reduce
//    Motion) stops the schedule outright so an idle / window-occluded host drops to
//    0 CPU while the static field stays on screen.
// 2. Geometry is CACHED per size (a one-slot memo). Per-frame work is only the
//    facet re-tint fills; no mesh rebuild, no hashing per frame (the Voice
//    original rebuilt the mesh every frame).
// 3. All randomness is the deterministic integer hash `nockerlFacetJitter01`,
//    IDENTICAL to Compose, so both platforms render the SAME field and nothing
//    ever re-rolls or shimmers.
// 4. Reduce Motion freezes the wave at phase 0: the field stays faceted (the
//    static grain remains) but stops drifting.
//
// Geometry/constants ported 1:1 from the canonical Compose implementation.

import SwiftUI

/// One triangular facet: corners, baked tonal offset, and diagonal position.
struct NockerlFacet {
    let a: CGPoint
    let b: CGPoint
    let c: CGPoint
    let staticDelta: Double
    let diagonalPos: Double
}

/// Deterministic pseudo-random in `0..<1` from two integer grid indices, the
/// SAME mixed-bit hash as Compose (`nockerlFacetJitter01`), so both platforms
/// render the identical field.
func nockerlFacetJitter01(_ x: Int32, _ y: Int32) -> Double {
    var h = x &* 374_761_393 &+ y &* 668_265_263
    h = (h ^ Int32(bitPattern: UInt32(bitPattern: h) >> 13)) &* 1_274_126_177
    h = h ^ Int32(bitPattern: UInt32(bitPattern: h) >> 16)
    return Double(h & 0xFFFFFF) / 16_777_216.0
}

/// Total luminance delta for a facet at animation `phase` (radians): baked
/// static delta + `amplitude * sin(phase + diagonalPos * 2π)`. Pure.
func nockerlFacetLuminanceDelta(
    staticDelta: Double,
    diagonalPos: Double,
    phase: Double,
    amplitude: Double = NockerlFacetedBackground.waveAmplitude
) -> Double {
    staticDelta + amplitude * sin(phase + diagonalPos * 2 * .pi)
}

/// Build the low-poly triangulation covering `size` (cells of `cell` points):
/// point grid with one extra ring beyond each edge, interior points nudged by
/// the deterministic hash, cells split into two triangles alternating the
/// diagonal by parity. Pure + unit-testable; identical geometry to Compose.
func nockerlBuildFacetField(
    size: CGSize,
    cell: CGFloat,
    jitterFraction: CGFloat = 0.34
) -> [NockerlFacet] {
    guard size.width > 0, size.height > 0, cell > 0 else { return [] }

    let cols = Int(size.width / cell) + 2
    let rows = Int(size.height / cell) + 2
    let jitter = cell * jitterFraction
    let diagSpan = max(size.width + size.height, 1)

    func point(_ col: Int, _ row: Int) -> CGPoint {
        let baseX = CGFloat(col) * cell
        let baseY = CGFloat(row) * cell
        let edge = col <= 0 || row <= 0 || col >= cols - 1 || row >= rows - 1
        if edge { return CGPoint(x: baseX, y: baseY) }
        let dx = (nockerlFacetJitter01(Int32(col), Int32(row)) - 0.5) * 2 * jitter
        let dy = (nockerlFacetJitter01(Int32(col + 101), Int32(row + 211)) - 0.5) * 2 * jitter
        return CGPoint(x: baseX + CGFloat(dx), y: baseY + CGFloat(dy))
    }

    func facet(
        _ tri: (CGPoint, CGPoint, CGPoint),
        _ col: Int,
        _ row: Int,
        _ half: Int
    ) -> NockerlFacet {
        let cx = (tri.0.x + tri.1.x + tri.2.x) / 3
        let cy = (tri.0.y + tri.1.y + tri.2.y) / 3
        let diagonalPos = min(max(Double((cx + cy) / diagSpan), 0), 1)
        let staticDelta =
            (nockerlFacetJitter01(Int32(col * 2 + half), Int32(row)) - 0.5) * 2
                * NockerlFacetedBackground.staticFacetJitter
        return NockerlFacet(
            a: tri.0, b: tri.1, c: tri.2,
            staticDelta: staticDelta, diagonalPos: diagonalPos
        )
    }

    var facets: [NockerlFacet] = []
    facets.reserveCapacity(cols * rows * 2)
    for row in 0..<(rows - 1) {
        for col in 0..<(cols - 1) {
            let tl = point(col, row)
            let tr = point(col + 1, row)
            let br = point(col + 1, row + 1)
            let bl = point(col, row + 1)
            let splitTlBr = (col + row) % 2 == 0
            let tri1 = splitTlBr ? (tl, tr, br) : (tl, tr, bl)
            let tri2 = splitTlBr ? (tl, br, bl) : (tr, br, bl)
            facets.append(facet(tri1, col, row, 0))
            facets.append(facet(tri2, col, row, 1))
        }
    }
    return facets
}

/// The faceted background fills its parent behind content. Reads `chatBg` +
/// `canvasEdge` from the palette, so it tracks the active theme.
public struct NockerlFacetedBackground: View {
    /// Nominal facet cell edge (Compose: 128dp), kept large so the mesh reads as
    /// a few big facets.
    static let cellSize: CGFloat = 128
    /// One full tone-wave cycle (seconds).
    static let wavePeriod: TimeInterval = 18
    /// Peak luminance swing the moving tone-wave adds to a facet.
    static let waveAmplitude: Double = 0.05
    /// Static per-facet luminance jitter range (±) baked at build.
    static let staticFacetJitter: Double = 0.022
    /// Alpha of the facet-edge hairline.
    static let edgeAlpha: Double = 0.05

    /// Freeze the tone-wave when `true`: the `TimelineView` schedule stops (0 CPU)
    /// while the faceted field stays on screen. Hosts set this when their window is
    /// occluded / backgrounded. Composed with Reduce Motion. Default `false`.
    private let paused: Bool

    /// Create the field.
    /// - Parameter paused: freeze the tone-wave (0 CPU) but keep the static field;
    ///   default `false` (the animating, pre-1.14.0 behavior).
    public init(paused: Bool = false) {
        self.paused = paused
    }

    public var body: some View {
        FieldBody(paused: paused)
    }

    /// One-slot geometry memo: the mesh rebuilds only when the size changes.
    /// A reference type mutated during draw, which is deliberate: the TimelineView
    /// is what drives invalidation, the cache never needs to.
    private final class FieldCache {
        var size: CGSize = .zero
        var facets: [NockerlFacet] = []

        func facets(for size: CGSize) -> [NockerlFacet] {
            if size != self.size {
                self.size = size
                facets = nockerlBuildFacetField(size: size, cell: NockerlFacetedBackground.cellSize)
            }
            return facets
        }
    }

    private struct FieldBody: View {
        let paused: Bool
        @Environment(\.colorScheme) private var colorScheme
        @Environment(\.accessibilityReduceMotion) private var reduceMotion
        @State private var cache = FieldCache()

        var body: some View {
            let palette = NockerlPalette.resolve(colorScheme)
            let base = palette.chatBg
            let edge = palette.canvasEdge.opacity(NockerlFacetedBackground.edgeAlpha)

            // 20fps cap (v1.14.0): `minimumInterval` throttles the schedule from the
            // display's up-to-120fps to 20fps: sub-perceptual for the ~18s sine, ~6x
            // cheaper. `paused` (host occlusion) or Reduce Motion stops it entirely.
            TimelineView(.animation(minimumInterval: 1.0 / 20.0, paused: reduceMotion || paused)) { timeline in
                Canvas { context, size in
                    // Continuous phase from the frame timestamp. A full sine
                    // cycle wraps seamlessly. Reduce Motion pins it to 0.
                    let phase: Double
                    if reduceMotion {
                        phase = 0
                    } else {
                        let t = timeline.date.timeIntervalSinceReferenceDate
                        let period = NockerlFacetedBackground.wavePeriod
                        phase = (t.truncatingRemainder(dividingBy: period) / period) * 2 * .pi
                    }

                    context.fill(Path(CGRect(origin: .zero, size: size)), with: .color(base))
                    for facet in cache.facets(for: size) {
                        let delta = nockerlFacetLuminanceDelta(
                            staticDelta: facet.staticDelta,
                            diagonalPos: facet.diagonalPos,
                            phase: phase
                        )
                        var path = Path()
                        path.move(to: facet.a)
                        path.addLine(to: facet.b)
                        path.addLine(to: facet.c)
                        path.closeSubpath()
                        context.fill(path, with: .color(NockerlContrast.shift(base, by: delta)))
                        context.stroke(path, with: .color(edge), lineWidth: 0.75)
                    }
                }
            }
        }
    }
}
